x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/utils.ts

```diff
@@ -193,6 +195,116 @@ export const getTupleDuplicateErrorsAndUniqueRules = (
+ * id in the provided action with the _id from the query result
+ * @param action
+ * @param esClient
+ * @returns
+ */
+export const swapActionIds = async (
+  action: Action,
+  savedObjectsClient: SavedObjectsClientContract
+): Promise<Action | Error> => {
+  try {
+    // TODO: might need to pass down namespace
+    // but for now it's blank to do local testing.
+    const search = createQuery('action', action.id, '');
+    const foundAction = await savedObjectsClient.find<Action>({
+      type: 'action',
+      search,
+      rootSearchFields: ['_id', 'originId'],
+    });
+
+    if (foundAction.saved_objects.length === 1) {
+      return { ...action, id: foundAction.saved_objects[0].id };
+    } else if (foundAction.saved_objects.length > 1) {
+      return new Error(
+        `action connector with originId: ${action.id} had conflicts. Please resolve these conflicts either in the file you are attempting to upload or resolve the conflicting action connector in Kibana.`
+      );
+    }
+  } catch (exc) {
+    return action;
+  }
+  return action;
+};
+
+/**
+ * In 8.0 all saved objects made in a non-default space will have their
+ * _id's regenerated. Security Solution rules have references to the
+ * actions SO id inside the 'actions' param.
+ * When users import these rules, we need to ensure any rule with
+ * an action that has an old, pre-8.0 id will need to be updated
+ * to point to the new _id for that action (alias_target_id)
+ *
+ * ex:
+ * import rule.ndjson:
+ * {
+ *   rule_id: 'myrule_id'
+ *   name: 'my favorite rule'
+ *   ...
+ *   actions:[{id: '1111-2222-3333-4444', group...}]
+ * }
+ *
+ * In 8.0 the 'id' field of this action is no longer a reference
+ * to the _id of the action (connector). Querying against the connector
+ * endpoint for this id will yield 0 results / 404.
+ *
+ * The solution: If we query the .kibana index for '1111-2222-3333-4444' as an originId,
+ * we should get the original connector back
+ * (with the new, migrated 8.0 _id of 'xxxx-yyyy-zzzz-0000') and can then replace
+ * '1111-2222-3333-4444' in the example above with 'xxxx-yyyy-zzzz-0000'
+ * And the rule will then import successfully.
+ * @param rules
+ * @param savedObjectsClient SO client exposing hidden 'actions' SO type
+ * @returns
+ */
+export const migrateLegacyActionsIds = async (
+  rules: PromiseFromStreams[],
+  savedObjectsClient: SavedObjectsClientContract
+): Promise<PromiseFromStreams[]> => {
+  const isImportRule = (r: unknown): r is ImportRulesSchemaDecoded => !(r instanceof Error);
+  return Promise.all(
+    rules.map(async (rule) => {
+      if (isImportRule(rule)) {
+        // can we swap the pre 8.0 action connector(s) id with the new,
+        // post-8.0 action id (swap the originId for the new _id?)
+        const newActions: Array<Action | Error> = await Promise.all(
+          rule.actions.map((action: Action) => swapActionIds(action, savedObjectsClient))
+        );
```

Comment:
You've got a `Promise.all` with a nested `Promise.all` here, and each one is firing off a `SOclient.find`; if someone is importing a lot of rules that have a lot of references to action connectors, you run the risk of exceeding Node's maximum call stack size.

I highly recommend using [`p-map`](https://www.npmjs.com/package/p-map) to limit concurrency and avoid this situation; we are doing this in the SOM import code here:

https://github.com/elastic/kibana/blob/e35be9d87c86a9340033193547a1b2c9a4be1bf7/src/core/server/saved_objects/import/lib/check_origin_conflicts.ts#L145-L150